File: app/src/main/java/com/nutomic/syncthingandroid/model/RemoteCompletion.java
package com.nutomic.syncthingandroid.model;

import android.util.Log;

import com.google.common.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.AbstractMap.SimpleEntry;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.text.TextUtils;

import com.nutomic.syncthingandroid.util.Util;

/**
 * This class caches remote folder and device synchronization
 * completion indicators defined in {@link RemoteCompletionInfo}
 * according to syncthing's REST "/completion" JSON result schema.
 * Completion model of syncthing's web UI is completion[deviceId][folderId]
 */
public class RemoteCompletion {

    private static final String TAG = "RemoteCompletion";

    private Boolean ENABLE_DEBUG_LOG = false;
    private boolean ENABLE_VERBOSE_LOG = false;

    HashMap<String, Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>>> mDeviceFolderMap =
        new HashMap<String, Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>>>();

    /**
     * Object that must be locked upon accessing mDeviceFolderMap.
     */
    private final Object mDeviceFolderMapLock = new Object();

    public RemoteCompletion(Boolean enableVerboseLog) {
        ENABLE_VERBOSE_LOG = enableVerboseLog;
    }

    /**
     * Removes a folder from the cache model.
     */
    private void removeFolder(String folderId) {
        synchronized(mDeviceFolderMapLock) {
            for (Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>> folderMapEntry : mDeviceFolderMap.values()) {
                HashMap<String, RemoteCompletionInfo> folderMap = folderMapEntry.getValue();
                if (folderMap.containsKey(folderId)) {
                    folderMap.remove(folderId);
                    break;
                }
            }
        }
    }

    /**
     * Updates device and folder information in the cache model
     * after a config update.
     */
    public void updateFromConfig(final List<Device> newDevices, final List<Folder> newFolders) {
        synchronized(mDeviceFolderMapLock) {
            HashMap<String, RemoteCompletionInfo> folderMap;

            // Handle devices that were removed from the config.
            List<String> removedDevices = new ArrayList<>();
            Boolean deviceFound;
            for (String deviceId : mDeviceFolderMap.keySet()) {
                deviceFound = false;
                for (Device device : newDevices) {
                    if (device.deviceID.equals(deviceId)) {
                        deviceFound = true;
                        break;
                    }
                }
                if (!deviceFound) {
                    removedDevices.add(deviceId);
                }
            }
            for (String deviceId : removedDevices) {
                LogV("updateFromConfig: Remove device '" + getShortenedDeviceId(deviceId) + "' from cache model");
                mDeviceFolderMap.remove(deviceId);
            }

            // Handle devices that were added to the config.
            for (Device device : newDevices) {
                if (!mDeviceFolderMap.containsKey(device.deviceID)) {
                    LogV("updateFromConfig: Add device '" + getShortenedDeviceId(device.deviceID) + "' to cache model");
                    mDeviceFolderMap.put(
                            device.deviceID,
                            new SimpleEntry(
                                    new Connection(),
                                    new HashMap<String, RemoteCompletionInfo>()
                            )
                    );
                }
            }

            // Handle folders that were removed from the config.
            List<String> removedFolders = new ArrayList<>();
            Boolean folderFound;
            for (Map.Entry<String, Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>>> device : mDeviceFolderMap.entrySet()) {
                //                            Map.Entry   HashMap    String
                for (String folderId : device.getValue().getValue().keySet()) {
                    folderFound = false;
                    for (Folder folder : newFolders) {
                        if (folder.id.equals(folderId)) {
                            folderFound = true;
                            break;
                        }
                    }
                    if (!folderFound) {
                        removedFolders.add(folderId);
                    }
                }
            }
            for (String folderId : removedFolders) {
                LogV("updateFromConfig: Remove folder '" + folderId + "' from cache model");
                removeFolder(folderId);
            }

            // Handle folders that were added to the config.
            for (Folder folder : newFolders) {
                for (Device device : newDevices) {
                    if (folder.getDevice(device.deviceID) != null) {
                        // folder is shared with device.
                        folderMap = mDeviceFolderMap.get(device.deviceID).getValue();
                        if (!folderMap.containsKey(folder.id)) {
                            LogV("updateFromConfig: Add folder '" + folder.id +
                                    "' shared with device '" + getShortenedDeviceId(device.deviceID) + "' to cache model.");
                            folderMap.put(folder.id, new RemoteCompletionInfo());
                        }
                    }
                }
            }
        }
    }

    /**
     * Accumulates per-folder completion values and calculates the average percentage
     * clamped to 0-100. Takes into account that Syncthing's WebUI considers remote
     * folders with 0% and 100% completion as up-to-date.
     */
    private static final class CompletionAccumulator {

        private int folderCount = 0;
        private double sumCompletion = 0;

        void accumulate(Iterable<RemoteCompletionInfo> completionInfos) {
            for (RemoteCompletionInfo completionInfo : completionInfos) {
                double folderCompletion = completionInfo.completion;
                if (folderCompletion < 0) {
                    folderCompletion = 0;
                } else if (folderCompletion > 100) {
                    folderCompletion = 100;
                }
                if (folderCompletion != 0 && folderCompletion != 100) {
                    sumCompletion += folderCompletion;
                    folderCount++;
                }
            }
        }

        int calculatePercentage() {
            if (folderCount == 0) {
                return 100;
            }
            int completion = (int) Math.floor(sumCompletion / folderCount);
            if (completion < 0) {
                completion = 0;
            } else if (completion > 100) {
                completion = 100;
            }
            return completion;
        }
    }

    /**
     * Calculates remote device sync completion percentage across all connected devices.
     * Returns "-1" if sync completion is not applicable.
     */
    public int getTotalDeviceCompletion() {
        synchronized(mDeviceFolderMapLock) {
            int connectedDeviceCount = 0;
            for (Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>> device : mDeviceFolderMap.values()) {
                if (device.getKey().connected) {
                    connectedDeviceCount++;
                }
            }
            if (connectedDeviceCount == 0) {
                return -1;
            }
            CompletionAccumulator accumulator = new CompletionAccumulator();
            for (Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>> device : mDeviceFolderMap.values()) {
                accumulator.accumulate(device.getValue().values());
            }
            return accumulator.calculatePercentage();
        }
    }

    /**
     * Calculates remote device sync completion percentage across all folders
     * shared with the device.
     */
    public int getDeviceCompletion(String deviceId) {
        synchronized(mDeviceFolderMapLock) {
            if (!mDeviceFolderMap.containsKey(deviceId)) {
                LogV("getDeviceCompletion: Cache miss for deviceId=[" + deviceId + "]");
                return 100;
            }

            HashMap<String, RemoteCompletionInfo> folderMap = mDeviceFolderMap.get(deviceId).getValue();
            if (folderMap != null) {
                CompletionAccumulator accumulator = new CompletionAccumulator();
                accumulator.accumulate(folderMap.values());
                return accumulator.calculatePercentage();
            }
            return 100;
        }
    }

    public double getDeviceNeedBytes(String deviceId) {
        synchronized(mDeviceFolderMapLock) {
            if (!mDeviceFolderMap.containsKey(deviceId)) {
                LogV("getDeviceNeedBytes: Cache miss for deviceId=[" + deviceId + "]");
                return 0;
            }

            double sumNeedBytes = 0;
            HashMap<String, RemoteCompletionInfo> folderMap = mDeviceFolderMap.get(deviceId).getValue();
            if (folderMap != null) {
                for (Map.Entry<String, RemoteCompletionInfo> folder : folderMap.entrySet()) {
                    sumNeedBytes += folder.getValue().needBytes;
                }
            }
            return sumNeedBytes;
        }
    }

    /**
     * Set completionInfo within the completion[deviceId][folderId] model.
     */
    public void setCompletionInfo(String deviceId, String folderId,
                                    final RemoteCompletionInfo completionInfo) {
        synchronized(mDeviceFolderMapLock) {
            // Add device parent node if it does not exist.
            if (!mDeviceFolderMap.containsKey(deviceId)) {
                mDeviceFolderMap.put(
                        deviceId,
                        new SimpleEntry(
                                new Connection(),
                                new HashMap<String, RemoteCompletionInfo>()
                        )
                );
            }
            LogV("setCompletionInfo: Storing " + completionInfo.completion + "% for folder \"" +
                    folderId + "\" at device \"" +
                    getShortenedDeviceId(deviceId) + "\".");
            // Add folder or update existing folder entry.
            mDeviceFolderMap.get(deviceId).getValue().put(folderId, completionInfo);
        }
    }

    /**
     * Returns remote device status.
     */
    public final Connection getDeviceStatus(final String deviceId) {
        synchronized(mDeviceFolderMapLock) {
            if (!mDeviceFolderMap.containsKey(deviceId)) {
                return new Connection();
            }
            //                                      Map.Entry     Connection
            Connection connection = mDeviceFolderMap.get(deviceId).getKey();
            return Util.deepCopy(connection, new TypeToken<Connection>(){}.getType());
        }
    }

    public int getOnlineDeviceCount() {
        synchronized(mDeviceFolderMapLock) {
            int onlineDeviceCount = 0;
            for (Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>> device : mDeviceFolderMap.values()) {
                if (device.getKey().connected) {
                    onlineDeviceCount++;
                }
            }
            return onlineDeviceCount;
        }
    }

    /**
     * Store remote device status for later when we need info for the UI.
     */
    public void setDeviceStatus(final String deviceId,
                                    final Connection connection) {
        synchronized(mDeviceFolderMapLock) {
            // Add device parent node if it does not exist.
            if (!mDeviceFolderMap.containsKey(deviceId)) {
                mDeviceFolderMap.put(
                        deviceId,
                        new SimpleEntry(
                                new Connection(),
                                new HashMap<String, RemoteCompletionInfo>()
                        )
                );
            }

            if (ENABLE_DEBUG_LOG) {
                Log.d(TAG, "setDeviceStatus: deviceId=\"" + deviceId + "\"" +
                        ", connected=" + Boolean.toString(connection.connected) +
                        ", paused=" + Boolean.toString(connection.paused)
                );
            }

            // Update device status information.
            Map.Entry updatedEntry = new SimpleEntry(
                    Util.deepCopy(connection, new TypeToken<Connection>(){}.getType()),
                    Util.deepCopy(
                            mDeviceFolderMap.get(deviceId).getValue(),
                            new TypeToken<HashMap<String, RemoteCompletionInfo>>(){}.getType()
                    )
            );
            mDeviceFolderMap.put(deviceId, updatedEntry);
        }
    }

    /**
     * Returns the first characters of the device ID for logging purposes.
     */
    public String getShortenedDeviceId(String deviceId) {
        return (TextUtils.isEmpty(deviceId) ? "" : deviceId.substring(0, 7));
    }

    private void LogV(String logMessage) {
        if (ENABLE_VERBOSE_LOG) {
            Log.v(TAG, logMessage);
        }
    }
}

File: app/src/main/java/com/nutomic/syncthingandroid/model/RemoteCompletion.kt
package com.nutomic.syncthingandroid.model

import android.util.Log
import com.google.gson.reflect.TypeToken
import com.nutomic.syncthingandroid.util.Util
import java.util.AbstractMap

/**
 * This class caches remote folder and device synchronization
 * completion indicators defined in {@link RemoteCompletionInfo}
 * according to syncthing's REST "/completion" JSON result schema.
 * Completion model of syncthing's web UI is completion[deviceId][folderId]
 */
class RemoteCompletion(enableVerboseLog: Boolean) {

    private var ENABLE_DEBUG_LOG = false
    private val ENABLE_VERBOSE_LOG = enableVerboseLog

    private val deviceFolderMap:
        MutableMap<String, Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>>> = HashMap()

    /**
     * Object that must be locked upon accessing deviceFolderMap.
     */
    private val deviceFolderMapLock = Any()

    /**
     * Removes a folder from the cache model.
     */
    private fun removeFolder(folderId: String) {
        synchronized(deviceFolderMapLock) {
            for (folderMapEntry in deviceFolderMap.values) {
                val folderMap = folderMapEntry.value
                if (folderMap.containsKey(folderId)) {
                    folderMap.remove(folderId)
                    break
                }
            }
        }
    }

    /**
     * Updates device and folder information in the cache model
     * after a config update.
     */
    fun updateFromConfig(newDevices: List<Device>, newFolders: List<Folder>) {
        synchronized(deviceFolderMapLock) {
            // Handle devices that were removed from the config.
            val removedDevices = ArrayList<String>()
            for (deviceId in deviceFolderMap.keys) {
                var deviceFound = false
                for (device in newDevices) {
                    if (device.deviceID == deviceId) {
                        deviceFound = true
                        break
                    }
                }
                if (!deviceFound) {
                    removedDevices.add(deviceId)
                }
            }
            for (deviceId in removedDevices) {
                logV("updateFromConfig: Remove device '${getShortenedDeviceId(deviceId)}' from cache model")
                deviceFolderMap.remove(deviceId)
            }

            // Handle devices that were added to the config.
            for (device in newDevices) {
                if (!deviceFolderMap.containsKey(device.deviceID)) {
                    logV("updateFromConfig: Add device '${getShortenedDeviceId(device.deviceID)}' to cache model")
                    deviceFolderMap[device.deviceID] = AbstractMap.SimpleEntry(
                        Connection(),
                        HashMap()
                    )
                }
            }

            // Handle folders that were removed from the config.
            val removedFolders = ArrayList<String>()
            for (device in deviceFolderMap.values) {
                for (folderId in device.value.keys) {
                    var folderFound = false
                    for (folder in newFolders) {
                        if (folder.id == folderId) {
                            folderFound = true
                            break
                        }
                    }
                    if (!folderFound) {
                        removedFolders.add(folderId)
                    }
                }
            }
            for (folderId in removedFolders) {
                logV("updateFromConfig: Remove folder '$folderId' from cache model")
                removeFolder(folderId)
            }

            // Handle folders that were added to the config.
            for (folder in newFolders) {
                for (device in newDevices) {
                    if (folder.getDevice(device.deviceID) != null) {
                        // folder is shared with device.
                        val folderMap = deviceFolderMap[device.deviceID]!!.value
                        if (!folderMap.containsKey(folder.id)) {
                            logV("updateFromConfig: Add folder '${folder.id}'" +
                                " shared with device '${getShortenedDeviceId(device.deviceID)}' to cache model.")
                            folderMap[folder.id] = RemoteCompletionInfo()
                        }
                    }
                }
            }
        }
    }

    /**
     * Accumulates per-folder completion values and calculates the average percentage
     * clamped to 0-100. Takes into account that Syncthing's WebUI considers remote
     * folders with 0% and 100% completion as up-to-date.
     */
    private class CompletionAccumulator {
        private var folderCount = 0
        private var sumCompletion = 0.0

        fun accumulate(completionInfos: Iterable<RemoteCompletionInfo>) {
            for (completionInfo in completionInfos) {
                var folderCompletion = completionInfo.completion
                if (folderCompletion < 0) {
                    folderCompletion = 0.0
                } else if (folderCompletion > 100) {
                    folderCompletion = 100.0
                }
                if (folderCompletion != 0.0 && folderCompletion != 100.0) {
                    sumCompletion += folderCompletion
                    folderCount++
                }
            }
        }

        fun calculatePercentage(): Int {
            if (folderCount == 0) {
                return 100
            }
            var completion = Math.floor(sumCompletion / folderCount).toInt()
            if (completion < 0) {
                completion = 0
            } else if (completion > 100) {
                completion = 100
            }
            return completion
        }
    }

    /**
     * Calculates remote device sync completion percentage across all connected devices.
     * Returns "-1" if sync completion is not applicable.
     */
    fun getTotalDeviceCompletion(): Int {
        synchronized(deviceFolderMapLock) {
            var connectedDeviceCount = 0
            for (device in deviceFolderMap.values) {
                if (device.key.connected) {
                    connectedDeviceCount++
                }
            }
            if (connectedDeviceCount == 0) {
                return -1
            }
            val accumulator = CompletionAccumulator()
            for (device in deviceFolderMap.values) {
                if (!device.key.connected) {
                    continue
                }
                accumulator.accumulate(device.value.values)
            }
            return accumulator.calculatePercentage()
        }
    }

    /**
     * Calculates remote device sync completion percentage across all folders
     * shared with the device.
     */
    fun getDeviceCompletion(deviceId: String): Int {
        synchronized(deviceFolderMapLock) {
            if (!deviceFolderMap.containsKey(deviceId)) {
                logV("getDeviceCompletion: Cache miss for deviceId=[$deviceId]")
                return 100
            }

            val folderMap = deviceFolderMap[deviceId]!!.value
            if (folderMap != null) {
                val accumulator = CompletionAccumulator()
                accumulator.accumulate(folderMap.values)
                return accumulator.calculatePercentage()
            }
            return 100
        }
    }

    fun getDeviceNeedBytes(deviceId: String): Double {
        synchronized(deviceFolderMapLock) {
            if (!deviceFolderMap.containsKey(deviceId)) {
                logV("getDeviceNeedBytes: Cache miss for deviceId=[$deviceId]")
                return 0.0
            }

            var sumNeedBytes = 0.0
            val folderMap = deviceFolderMap[deviceId]!!.value
            if (folderMap != null) {
                for (folder in folderMap.values) {
                    sumNeedBytes += folder.needBytes
                }
            }
            return sumNeedBytes
        }
    }

    /**
     * Set completionInfo within the completion[deviceId][folderId] model.
     */
    fun setCompletionInfo(deviceId: String, folderId: String, completionInfo: RemoteCompletionInfo) {
        synchronized(deviceFolderMapLock) {
            // Add device parent node if it does not exist.
            if (!deviceFolderMap.containsKey(deviceId)) {
                deviceFolderMap[deviceId] = AbstractMap.SimpleEntry(
                    Connection(),
                    HashMap()
                )
            }
            logV("setCompletionInfo: Storing ${completionInfo.completion}% for folder \"" +
                "$folderId\" at device \"" +
                "${getShortenedDeviceId(deviceId)}\".")
            // Add folder or update existing folder entry.
            deviceFolderMap[deviceId]!!.value[folderId] = completionInfo
        }
    }

    /**
     * Returns remote device status.
     */
    fun getDeviceStatus(deviceId: String): Connection {
        synchronized(deviceFolderMapLock) {
            if (!deviceFolderMap.containsKey(deviceId)) {
                return Connection()
            }
            val connection = deviceFolderMap[deviceId]!!.key
            return Util.deepCopy(connection, object : TypeToken<Connection>() {}.type)
        }
    }

    fun getOnlineDeviceCount(): Int {
        synchronized(deviceFolderMapLock) {
            var onlineDeviceCount = 0
            for (device in deviceFolderMap.values) {
                if (device.key.connected) {
                    onlineDeviceCount++
                }
            }
            return onlineDeviceCount
        }
    }

    /**
     * Store remote device status for later when we need info for the UI.
     */
    fun setDeviceStatus(deviceId: String, connection: Connection) {
        synchronized(deviceFolderMapLock) {
            // Add device parent node if it does not exist.
            if (!deviceFolderMap.containsKey(deviceId)) {
                deviceFolderMap[deviceId] = AbstractMap.SimpleEntry(
                    Connection(),
                    HashMap()
                )
            }

            if (ENABLE_DEBUG_LOG) {
                Log.d(TAG, "setDeviceStatus: deviceId=\"$deviceId\"" +
                    ", connected=${connection.connected}" +
                    ", paused=${connection.paused}")
            }

            // Update device status information.
            val updatedEntry: Map.Entry<Connection, HashMap<String, RemoteCompletionInfo>> =
                AbstractMap.SimpleEntry(
                    Util.deepCopy(connection, object : TypeToken<Connection>() {}.type),
                    Util.deepCopy(
                        deviceFolderMap[deviceId]!!.value,
                        object : TypeToken<HashMap<String, RemoteCompletionInfo>>() {}.type
                    )
                )
            deviceFolderMap[deviceId] = updatedEntry
        }
    }

    /**
     * Returns the first characters of the device ID for logging purposes.
     */
    fun getShortenedDeviceId(deviceId: String): String {
        return if (deviceId.isEmpty()) "" else deviceId.substring(0, 7)
    }

    private fun logV(logMessage: String) {
        if (ENABLE_VERBOSE_LOG) {
            Log.v(TAG, logMessage)
        }
    }

    companion object {
        private const val TAG = "RemoteCompletion"
    }
}
